crates/recoverable/src/lib.rs

```diff
@@ -0,0 +1,546 @@
+// Copyright (c) Microsoft Corporation.
+
+//! Recovery metadata and classification for resilience patterns.
+//!
+//! This crate provides types for classifying conditions based on their **recoverability state**,
+//! enabling consistent recovery behavior across different error types and resilience middleware.
+//!
+//! The recovery metadata describes whether recovering from an operation might help, not whether
+//! the operation succeeded or failed. Both successful operations and permanent failures
+//! should use [`Recovery::never()`] since recovery won't change the outcome.
+//!
+//! # Core Types
+//!
+//! - [`Recovery`]: Classifies conditions as recoverable (transient) or non-recoverable (permanent/successful).
+//! - [`Recover`]: A trait for types that can determine their recoverability.
+//! - [`RecoveryKind`]: An enum representing the kind of recovery that can be attempted.
+//!
+//! # Examples
+//!
+//! ```rust
+//! use recoverable::{Recover, Recovery, RecoveryKind};
+//!
+//! #[derive(Debug)]
+//! enum DatabaseError {
+//!     ConnectionTimeout,
+//!     InvalidCredentials,
+//!     TableNotFound,
+//! }
+//!
+//! impl Recover for DatabaseError {
+//!     fn recovery(&self) -> Recovery {
+//!         match self {
+//!             // Transient failure - might succeed if retried
+//!             DatabaseError::ConnectionTimeout => Recovery::retry(),
+//!             // Permanent failures - retrying won't help
+//!             DatabaseError::InvalidCredentials => Recovery::never(),
+//!             DatabaseError::TableNotFound => Recovery::never(),
+//!         }
+//!     }
+//! }
+//!
+//! let error = DatabaseError::ConnectionTimeout;
+//! assert_eq!(error.recovery().kind(), RecoveryKind::Retry);
+//!
+//! // For successful operations, also use never() since retry is unnecessary
+//! let success_result: Result<(), DatabaseError> = Ok(());
+//! // If we had a wrapper type for success, it would also return Recovery::never()
+//! ```
+
+use std::fmt::{Display, Formatter};
+use std::time::Duration;
+
+/// Represents the recoverability metadata associated with an operation or condition.
+///
+/// This type describes how the operation can be recovered from, if at all. It provides
+/// various ways to create recovery metadata for different scenarios, such as unknown conditions,
+/// permanent failures, transient failures, and service unavailability.
+///
+/// # Examples
+///
+/// ```rust
+/// use recoverable::{Recovery, RecoveryKind};
+///
+/// let recovery = Recovery::retry();
+/// assert_eq!(recovery.kind(), RecoveryKind::Retry);
+/// ```
+#[derive(Debug, PartialEq, Clone)]
+#[non_exhaustive]
+pub struct Recovery {
+    kind: RecoveryKind,
+    delay: Option<Duration>,
```

Comment:
Hmmm. Wouldn't this be better as a parameter to the RecoveryKind::Retry enum? Either that, or add RecoveryKind::RetryAfter